metapool/plate.py

```diff
@@ -214,3 +217,198 @@ def _plate_position(well):
+        :param plate_df: processing-plate Pandas dataframe
+        :param threshold: Upper gDNA concentration threshold.
+        :return: Pandas DataFrame
+               If dilution was needed, gDNA concentrations will be adjusted.
+               'Diluted' column will be added and contents set to 'True'.
+    """
+    if 'Diluted' in plate_df.columns:
+        # function has already been applied to data
+        warnings.warn('Dilution operation was already performed')
+        return plate_df
+
+    # return a copy of the input data. Do not overwrite the input data by
+    # default.
+    df = plate_df.copy()
+    df['Diluted'] = True
+
+    df['Project Plate'] = df['Project Plate'].astype(str) + '_diluted'
+    df['Compressed Plate Name'] = df['Compressed Plate Name'] + '_dilution'
+    df['Sample DNA Concentration'] = df['Sample DNA Concentration'] / 10
+
+    # Picking diluted samples for normalization
+    diluteds = (df.loc[df['Sample DNA Concentration'] > (threshold / 10)])
+    undiluteds = plate_df.loc[~plate_df['Sample'].isin(diluteds['Sample'])]
+
+    return pd.concat([diluteds, undiluteds]).sort_index()
+
+
+def requires_dilution(plate_df, threshold=15, tolerance=0.05):
+    """
+        Reads a plate_df and determines whether or not it needs diluting.
+        :param plate_df: processing-plate Pandas dataframe
+        :param threshold: Upper gDNA concentration threshold.
+        :param tolerance: Percentage (0.0-1.0) allowed outside threshold.
+        :return: True if plate requires dilution. False otherwise.
+        """
+    # generate a Series of Booleans, based on whether each concentration falls
+    # within the threshold or not.
+    within_threshold = plate_df['Sample DNA Concentration'] > threshold
+    pct_of_total = plate_df.loc[within_threshold].shape[0] / plate_df.shape[0]
+
+    if pct_of_total > tolerance:
+        return True
+
+    return False
+
+
+def find_threshold(concentrations, labels):
+    """
+    Returns value that best separates concentration data into labels.
+    :param concentrations: Pandas series w/concentrations.
+    :param labels: Pandas series w/categorical labels.
+    :return: Threshold value.
+    """
+    # Training model
+    log_reg = LogisticRegression()
+    log_reg.fit(np.array(concentrations).reshape(-1, 1), np.array(labels))
+
+    # Making label predictions
+    predictions = pd.DataFrame(
+        [log_reg.predict(np.array(concentrations).reshape(-1, 1)),
+         np.array(labels)]).transpose().rename(
+        columns={0: 'predicted', 1: 'ground_truth'})
+
+    # Appending concentration values to predictions dataframe and sorting.
+    predictions['concentration'] = concentrations
+    predictions = predictions.sort_values('concentration').reset_index()
+
+    for i in predictions.index:
+        # Finds the first sample predicted as non-blank.
+        if predictions.loc[i, 'predicted'] == False:    # noqa: E712
```

Comment:
what about? `if not if predictions.loc[i, 'predicted']`?

Reply:
Assuming you mean 'if not' and not 'if not if' then sure.